packages/fuels-contract/src/contract.rs

```diff
@@ -410,17 +430,125 @@ where
     }
 
     /// Create a CallResponse from call receipts
-    pub fn get_response(&self, receipts: Vec<Receipt>) -> Result<CallResponse<D>, Error> {
+    pub fn get_response(&self, mut receipts: Vec<Receipt>) -> Result<CallResponse<D>, Error> {
         match self.contract_call.output_param.as_ref() {
             None => Ok(CallResponse::new(D::from_token(Token::Unit)?, receipts)),
             Some(param_type) => {
-                let (token, receipts) = ContractCall::get_decoded_output(param_type, receipts)?;
+                let token = ContractCall::get_decoded_output(param_type, &mut receipts)?;
                 Ok(CallResponse::new(D::from_token(token)?, receipts))
             }
         }
     }
 }
 
+#[derive(Debug)]
+#[must_use = "contract calls do nothing unless you `call` them"]
+/// Helper that handles bundling multiple calls into a single transaction
+pub struct MultiContractCallHandler {
+    pub contract_calls: Option<Vec<ContractCall>>,
+    pub tx_parameters: TxParameters,
+    pub wallet: LocalWallet,
+    pub fuel_client: FuelClient,
+}
+
+impl MultiContractCallHandler {
+    pub fn new(wallet: LocalWallet) -> Self {
+        Self {
+            contract_calls: None,
+            tx_parameters: TxParameters::default(),
+            fuel_client: wallet.get_provider().unwrap().client.clone(),
+            wallet,
+        }
+    }
+
+    /// Adds a contract call to be bundled in the transaction
+    /// Note that this is a builder method
+    pub fn add_call<D: Tokenizable>(&mut self, call_handler: ContractCallHandler<D>) -> &mut Self {
+        match self.contract_calls.as_mut() {
+            Some(c) => c.push(call_handler.contract_call),
+            None => self.contract_calls = Some(vec![call_handler.contract_call]),
+        }
+        self
+    }
+
+    /// Clears all added contract calls
+    /// Note that this is a builder method
+    pub fn clear_calls<D: Tokenizable>(&mut self) -> &mut Self {
+        if let Some(calls) = self.contract_calls.as_mut() {
+            calls.clear()
+        }
+        self
+    }
+
+    /// Sets the transaction parameters for a given transaction.
+    /// Note that this is a builder method
+    pub fn tx_params(&mut self, params: TxParameters) -> &mut Self {
+        self.tx_parameters = params;
+        self
+    }
+
+    /// Returns the script that executes the contract calls
+    pub async fn get_script(&self) -> Script {
+        Script::from_contract_calls(
+            self.contract_calls.as_ref().unwrap().iter().collect(),
+            &self.tx_parameters,
+            &self.wallet,
+        )
+        .await
+    }
+
+    /// Call contract methods on the node, in a state-modifying manner.
+    pub async fn call<D: Tokenizable + Debug>(&self) -> Result<MultiCallResponse<D>, Error> {
```

Comment:
It may look simple, but this line took us days of debugging and redesign to get right 😂

Reply:
Also the trick with bundling the output tokens of all the separate calls into a `Token::Tupple` so that the generic `D` effectively covers multiple types to avoid the need for dynamic dispatch